cloudinit/sources/DataSourceOVF.py

```diff
@@ -684,4 +741,98 @@ def _raise_error_status(prefix, error, event, config_file, conf):
     util.del_dir(os.path.dirname(config_file))
     raise error
 
+
+def load_cloudinit_data(metaPath, userPath):
+    """
+    Load the cloud-init meta data, user data, cfg and network from the
+    given files
+    """
+    LOG.debug('load meta data from: %s: user data from: %s',
+              metaPath, userPath)
+    md = {}
+    cfg = {}
+    ud = None
+    network = None
+
+    md = load(util.load_file(metaPath).replace("\r", ""))
+
+    if 'network' in md:
+        network = md['network']
+        del md['network']
+
+    if userPath:
+        ud = util.load_file(userPath).replace("\r", "")
+    return md, ud, cfg, network
+
+
+class LoadError(Exception):
+    pass
+
+
+class FileNotFound(Exception):
+    pass
+
+
+def load(data):
+    '''
+    first attempts to unmarshal the provided data as JSON, and if
+    that fails then attempts to unmarshal the data as YAML. If data is
+    None then a new dictionary is returned.
+    '''
+    if not data:
+        return {}
+    try:
+        return json.loads(data)
+    except Exception:
+        try:
+            return safeyaml.load(data)
+        except Exception as e:
+            raise LoadError("Error parsing the meta data") from e
+
+
+def collect_imc_files(cust_conf):
+    '''
+    wait the "VMware Tools" daemon to copy all the customization files to
+    /var/run/vmware-imc directory.
+    - meta data:
+      mandatory if customization specification is raw cloud-init data
+    - user data:
+      optional if customization specification is raw cloud-init data
+    - nics.txt:
+      mandatory if customization specification is traditional one
+    '''
+    metaPath = None
+    userPath = None
+    nicsPath = None
+    metaDataFile = cust_conf.meta_data_name
+    if metaDataFile:
+        metaPath = util.log_time(
+            logfunc=LOG.debug,
+            msg="waiting for meta data file",
+            func=wait_for_imc_file,
+            args=(metaDataFile, 10))
+        if not metaPath:
+            raise FileNotFound("cloud-init meta data file is not found")
+
+        userDataFile = cust_conf.user_data_name
+        if userDataFile:
+            userPath = util.log_time(
```

Comment:
is there a reason to wait here? If metadatafile is present, can we just assume that it was written last? and thus userDataFile will already exist? 

It seems at least there is a race condition otherwise where we could read the metadataFile, and then not have a userDataFile for 10 seconds, but then one appear as soon as we stop looking.

It would be nice for there to be a guarantee from the platform in some way so we didn't have such a race.

Perhaps a third file "files.json" could list the files that were present, and then we could wait for that explicit list? or some other way so we don't potentially go on without finding a file that would have been there later.

Reply:
Thanks, I will add these files as a list and call util.wait_for_files(). Also, we will ensure meta data is not last file to be copied.

Reply:
The vmware platform will ensure cust.cfg is copied as the last file， so we could assume these 3 files are copied there, here we needn't to wait for any files. I will change the codes accordingly.